sherlock/internal/api/sherlock/service_alerts_procedures_v3_sync.go

```diff
@@ -0,0 +1,197 @@
+		return
+	}
+	// set GCS client
+	var gcsClient, gc_err = google_bucket.InitializeStorageClient(ctx)
+	if gc_err != nil {
+		return
+	}
+	// get alerts blob
+	alert_json_blob, read_err := gcsClient.GetBlob(ctx, *gcs_bucket, "alerts.json")
+	if read_err != nil {
+		errors.AbortRequest(ctx, fmt.Errorf("blob not found: %v", read_err))
+		return
+	}
+	// read data from blob & parse to JSON
+	json_data := getFileJson(ctx, gcsClient, alert_json_blob)
+	// diff alerts from db and json data for currently live service alerts
+	add, remove, update := compareAlerts(ctx, json_data, alerts)
+	// if any changes then re-create file and upload to gcs bucket
+	if add != nil || remove != nil || update != nil {
+		// convert model data for service alerts to json formatted bytes to upload
+		json_bytes, err := createSvcAlertJsonData(ctx, alerts)
+		if err != nil {
+			errors.AbortRequest(ctx, fmt.Errorf("issue encountered creating json payload: %v", err))
+			return
+		}
+		// Upload file to bucket w/ latest info
+		if err = gcsClient.WriteBlob(ctx, *gcs_bucket, "alerts.json", json_bytes); err != nil {
+			errors.AbortRequest(ctx, fmt.Errorf("error writing updated alerts.json file: %v", err))
+			return
+		}
+	}
+}
+
+func getAlerts(ctx *gin.Context, request ServiceAlertV3SyncRequest, db *gorm.DB) ([]models.ServiceAlert, *string) {
+	var env_result models.Environment
+	if request.OnEnvironment != nil {
+		// match env so that we can get gcs bucket
+		environmentQuery, err := environmentModelFromSelector(*request.OnEnvironment)
+		if err != nil {
+			errors.AbortRequest(ctx, fmt.Errorf("error parsing environment selector '%s': %w", *request.OnEnvironment, err))
+			return nil, nil
+		}
+		if err = db.Where(&environmentQuery).Select("id").First(&env_result).Error; err != nil {
+			errors.AbortRequest(ctx, fmt.Errorf("error fetching environment '%v'", err))
+			return nil, nil
+		}
+	}
+	var activeAlerts []models.ServiceAlert
+	// Only return service alerts that haven't been deleted for this environment
+	if err := db.Model(&models.ServiceAlert{}).Where("DeletedAt = '' AND OnEnvironmentID = '%v'", env_result.ID).Find(&activeAlerts).Error; err != nil {
+		errors.AbortRequest(ctx, fmt.Errorf("(%s) error querying for Service Alerts: %w", errors.InternalServerError, err))
+		return nil, nil
+	}
+	return activeAlerts, env_result.ServiceBannerBucket
+
+}
+
+func getFileJson(ctx *gin.Context, gcs_client *google_bucket.GcsClientActual, blob *storage.ObjectAttrs) []ServiceAlertJsonData {
+	byte_data, read_err := gcs_client.ReadBlob(ctx, blob)
+	if read_err != nil {
+		// handle error
+		return nil
+	}
+	var json_data []ServiceAlertJsonData
+	err := json.Unmarshal(byte_data, &json_data)
+	if err != nil {
+		// handle error
+		return nil
+	}
+	return json_data
+}
+
+func compareAlerts(ctx *gin.Context, json_slice []ServiceAlertJsonData, db_alerts []models.ServiceAlert) ([]ServiceAlertJsonData, []models.ServiceAlert, []ServiceAlertJsonData) {
```

Comment:
You might want to add some comments here to talk about the logical actions, this looks like a pretty complex method and I'm not entirely clear why `toRemove` is different from `toAdd` and `toUpdate`